index.rst

```diff
@@ -1,25 +1,429 @@
+  As a result, poorly written tests may fail due to timing issues.
+
+I converted three packages to work with Kafka ts_salobj: ``ts_atwhitelight``, ``ts_scriptqueue`` and ``ts_watcher``.
+The first is a simple CSC.
+The latter two have the most complex use of ts_salobj that I know of (except for ``ts_salkafka``, which is not needed with Kafkfa ts_salobj).
+I only touched the code, not the build files.
+Here is what I found:
+
+ts_atwhitelight
+~~~~~~~~~~~~~~~
+
+No changes required.
+
+ts_watcher
+~~~~~~~~~~
+
+Changes required:
+
+* Move ``super().start()`` to the top of ``WatcherCsc.start``.
+
+* Some unit tests create ``SalIinfo`` instance, in unit tests in order to write mock data from other CSCs.
+  Call ``await start()`` on those instances.
+
+ts_scriptqueue
+~~~~~~~~~~~~~~
+
+Changes required:
+
+* Move ``super().start()`` to the top of ``ScriptQueue.start``.
+
+* Change ``ScriptID`` to ``private_index``.
+  In one case, delete a line that was needed to handle the fact that the field name ``ScriptID`` did not start with ``private_``.
+
+* Change one call from ``message.get_vars()`` to ``vars(message)``.
+
+* Fix one instance of writing a bytes array as a string, by encoding the bytes array.
+
+* Fix one test that failed due to tests timing issues.
+  The fix is backward compatible with DDS ts_salobj.
+
+Build Files
+-----------
+
+All packages will need build files updated, including:
+
+* Require Kafka-related packages instead of dds-related packages.
+
+* Stop building IDL files in Jenkinsfile.
+  Kafka ts_salobj parses our ts_xml topic schema files directly.
+
+* Update our test configuration to use ``pytest-docker`` to start and stop ``kafka-aggregator`` so unit tests can run.
+
+Appendix: Technical Details of the Kafka Implementation
+=======================================================
+
+Schemas
+-------
+
+Confluent Schema Registry supports `three schema formats <https://www.confluent.io/blog/confluent-platform-now-supports-protobuf-json-schema-custom-formats/>`_: Avro, Google protobuf, and json schema.
+
+Avro is the standard solution.
+In addition, our code that feeds Kafka data to the EFD uses Avro, and would not be easy to switch.
+So for my Kafka ts_salobj prototype I used Avro.
+Nonetheless I looked into the other options before making that decision, so I document what I found here.
+
+Support for Google protobuf, and json schema is recent.
+Information is very sparse about how Kafka uses json schema.
+For example: does Kafka fill in default values? the standard json schema validators do not.
+
+Here are two comparisons between Avro and protobuf: a `summary <https://blog.softwaremill.com/the-best-serialization-strategy-for-event-sourcing-9321c299632b>`_ that likes both, and a `detailed look <detailed look>`_, including schema evolution.
+It looks to me as if protobuf is very similar to Avro, with similar advantages and disadvantages.
+I see no strong reason to prefer protobuf over Avro.
```

Comment:
> Therefore, there seems to be no strong reason to prefer protobuf over Avro.

Reply:
I over-used "similar" so I'm trying this, instead:

Google protobuf has similar advantages and disadvantages to Avro.
Therefore we see no strong reason to prefer protobuf over Avro.